theano/gpuarray/sort.py

```diff
@@ -0,0 +1,314 @@
+        return kernels
+
+    def c_code(self, node, nodename, inps, outs, sub):
+        if node.inputs[0].type.context.kind != b'cuda':
+            raise NotImplementedError(
+                '%s: We only have CUDA '
+                'implementation so far.' % self.__class__.__name__)
+        x, k = inps
+        inp_dtc = ga.dtype_to_typecode(node.inputs[0].dtype)
+        if not self.return_indices:
+            yv, = outs
+        elif self.return_values:
+            yv, yi = outs
+        else:
+            yi, = outs
+        out_dtype_s = self.idx_dtype
+        out_dtc = ga.dtype_to_typecode(out_dtype_s)
+        fail = sub['fail']
+        ctx = sub['params']
+        k_dtype = node.inputs[1].type.dtype_specs()[1]
+        MAX_TPB = 1024  # max threads per block
+        WARP_SIZE = 32
+
+        ndim = node.inputs[0].ndim
+        reordered_axes = list(range(ndim))
+        axis = self.axis % ndim
+        del(reordered_axes[axis])
+        reordered_axes = [axis] + reordered_axes
+        dims = ''.join('(void*)(dims+%d), ' % i for i in reordered_axes[1:])
+        prep_output = ''
+        if self.return_values:
+            def_dvstrides = 'const ssize_t *dvstrides = PyGpuArray_STRIDES(%s)' % yv
+            params_dv = '(void*)(%s->ga.data),\n' % yv
+            params_dv += ''.join('(void*)(dvstrides+%d), ' % i for i in reordered_axes)
+            prep_output += '''
+    if (0 != theano_prep_output(
+        &%(yv)s, %(ndim)d, odims,
+        %(inp_dtc)s, GA_C_ORDER, %(ctx)s)) {
+        %(fail)s;
+    }\n''' % locals()
+        else:
+            def_dvstrides = params_dv = ''
+
+        if self.return_indices:
+            def_distrides = 'const ssize_t *distrides = PyGpuArray_STRIDES(%s)' % yi
+            params_di = '(void*)(%s->ga.data),\n' % yi
+            params_di += ''.join('(void*)(distrides+%d), ' % i for i in reordered_axes)
+            prep_output += '''
+    if (0 != theano_prep_output(
+        &%(yi)s, %(ndim)d, odims,
+        %(out_dtc)s, GA_C_ORDER, %(ctx)s)) {
+        %(fail)s;
+    }\n''' % locals()
+        else:
+            def_distrides = params_di = ''
+        sstrides = ', '.join('(void*)(sstrides+%d)' % i for i in reordered_axes)
+        code = '''
+{
+    const ssize_t k_ = ((%(k_dtype)s*)(PyArray_DATA(%(k)s)))[0];
+    const size_t *dims = PyGpuArray_DIMS(%(x)s);
+    size_t odims[%(ndim)d];
+    for (int i=0; i<%(ndim)d; i++)
+        odims[i] = dims[i];
+
+    odims[%(axis)d] = k_>=0 ? k_ : -k_;
+
+    if (0 == odims[%(axis)d]) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "topk: kth must not be zero");
+        %(fail)s;
+    } else if (dims[%(axis)d] < odims[%(axis)d]) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "topk: kth cannot be larger than the size of specified axis %(axis)d");
+        %(fail)s;
+    }
+    %(prep_output)s
+
+    size_t blk[6];
+    size_t *grd = blk+3;
```

Comment:
This is just obscuring intent.  Please declare two arrays.

Also you never use anything other than the first element so you can declare them as size 1 (or as scalars).